platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/impl/ApplicationServiceBean.java

```diff
@@ -362,96 +366,84 @@ public void removeApplication(String appName) {
+    Set<MetaTypeInformation> metatypeInformation =
+        (metatypeService == null)
+            ? Collections.emptySet()
+            : Arrays.stream(getContext().getBundles())
+                .filter(b -> bundleLocations.contains(b.getLocation()))
+                .map(metatypeService::getMetaTypeInformation)
+                .collect(Collectors.toSet());
+
+    return services
+        .stream()
+        .filter(
+            service ->
+                hasBundleLocation(service, bundleLocations)
+                    || hasMetatypesForService(service, metatypeInformation))
+        .collect(Collectors.toList());
+  }
 
-    return returnValues;
+  /**
+   * Checks to see if a given service configuration has a bundle location in the set provided.
+   *
+   * @param service the service with a location to check.
+   * @param bundleLocations a set of bundle locations to check.
+   * @return true if the service's location is in bundleLocations, false otherwise.
+   */
+  private boolean hasBundleLocation(Map<String, Object> service, Set<String> bundleLocations) {
+    return Stream.of(service)
+        .filter(m -> m.containsKey("configurations"))
+        .map(m -> m.get("configurations"))
+        .map(List.class::cast)
+        .flatMap(List::stream)
+        .map(Map.class::cast)
+        .filter(m -> m.containsKey("bundle_location"))
+        .map(m -> m.get("bundle_location"))
+        .map(String.class::cast)
+        .anyMatch(bundleLocations::contains);
   }
 
   /**
    * Checks to see if there are any metatypes out there for a particular service.
    *
-   * @param metatypeInformations - Where we'll look for metatypes that match our service from.
    * @param service - our service we want metatypes for.
+   * @param metatypeInformations - Where we'll look for metatypes that match our service from.
    * @return true if there is, and the service should be added, or false if it shouldn't be.
    */
-  private boolean checkForMetaTypesForService(
-      Set<MetaTypeInformation> metatypeInformations, Map<String, Object> service) {
+  private boolean hasMetatypesForService(
+      Map<String, Object> service, Set<MetaTypeInformation> metatypeInformations) {
     String id = (String) service.get("id");
-    boolean ifFactory = (Boolean) service.get("factory");
-    if (ifFactory) {
-      for (MetaTypeInformation information : metatypeInformations) {
-        if (information != null) {
-          for (String pid : information.getFactoryPids()) {
-            if (pid.equals(id)) {
-              return true;
-            }
-          }
-        }
-      }
-    } else {
-      for (MetaTypeInformation information : metatypeInformations) {
-        if (information != null) {
-          for (String pid : information.getPids()) {
-            if (pid.equals(id)) {
-              return true;
-            }
-          }
-        }
-      }
+    Boolean isFactory = (Boolean) service.get("factory");
+
+    if (id == null || isFactory == null) {
```

Comment:
✏️  Move the `if (id == null)` before defining `isFactory` to avoid getting it if not needed